lib/iris/analysis/stats.py

```diff
@@ -79,90 +82,93 @@ def pearsonr(
-        else:
-            mask_cube = 1.0
-        if ma.is_masked(cube_1.data):
-            # Take a slice to avoid unnecessary broadcasting of cube_2.
-            slice_coords = [
-                dim_coords_1[i]
-                for i in range(cube_1.ndim)
-                if dim_coords_1[i] not in common_dim_coords
-                and np.array_equal(
-                    cube_1.data.mask.any(axis=i), cube_1.data.mask.all(axis=i)
-                )
-            ]
-            cube_1_slice = next(cube_1.slices_over(slice_coords))
-            mask_cube = _ones_like(cube_1_slice) * mask_cube
-        # Apply common mask to data.
-        if isinstance(mask_cube, iris.cube.Cube):
-            cube_1 = cube_1 * mask_cube
-            cube_2 = mask_cube * cube_2
-            dim_coords_2 = [coord.name() for coord in cube_2.dim_coords]
-
-    # Broadcast weights to shape of cubes if necessary.
-    if weights is None or cube_1.shape == smaller_shape:
-        weights_1 = weights
-        weights_2 = weights
+        mask_1 = al.ma.getmaskarray(array_1)
+        if al is np:
+            # Reduce all invariant dimensions of mask_1 to length 1.  This avoids
+            # unnecessary broadcasting of array_2.
+            index = tuple(
+                slice(0, 1)
+                if np.array_equal(mask_1.any(axis=dim), mask_1.all(axis=dim))
+                else slice(None)
+                for dim in range(mask_1.ndim)
+            )
+            mask_1 = mask_1[index]
+
+        array_2 = _mask_array(array_2, mask_1)
+        array_1 = _mask_array(array_1, al.ma.getmaskarray(array_2))
+
+    # Broadcast weights to shape of arrays if necessary.
+    if weights is None:
+        weights_1 = weights_2 = None
     else:
         if weights.shape != smaller_shape:
-            raise ValueError(
-                "weights array should have dimensions {}".format(smaller_shape)
-            )
+            msg = f"weights array should have dimensions {smaller_shape}"
+            raise ValueError(msg)
+
+        if resolver.reorder_src_dims is not None:
+            # Apply same transposition as was done to cube_2 within Resolve.
+            weights = weights.transpose(resolver.reorder_src_dims)
+
+        # Reshape to add in any length-1 dimensions needed for broadcasting.
+        weights = weights.reshape(cube_2.shape)
 
-        dims_1_common = [
-            i for i in range(cube_1.ndim) if dim_coords_1[i] in common_dim_coords
-        ]
-        weights_1 = broadcast_to_shape(weights, cube_1.shape, dims_1_common)
-        if cube_2.shape != smaller_shape:
-            dims_2_common = [
-                i for i in range(cube_2.ndim) if dim_coords_2[i] in common_dim_coords
-            ]
-            weights_2 = broadcast_to_shape(weights, cube_2.shape, dims_2_common)
-        else:
-            weights_2 = weights
+        weights_2 = np.broadcast_to(weights, array_2.shape)
+        weights_1 = np.broadcast_to(weights, array_1.shape)
```

Comment:
Is there any merit in:

```python
weights_cube = Cube(weights)
weights_resolver = Resolve(cube_1, weights_cube)
weights_1 = weights_resolver.lhs_cube_resolved
weights_2 = weights_resolver.rhs_cube_resolved
```

In my head this would avoid needing to get `pearsonr()` to track the logic of `Resolve()`, plus make the code simpler.

But I could be missing an important reason against this.

Reply:
It would need to be

```python
weights_cube = smaller_cube.copy(weights)
...etc...
```
so that `Resolve` has all the metadata it needs to figure out the mappings.  The downside is that we'd be going through all `Resolve`'s logic twice, which could be inefficient.  Maybe something to look at once we have the benchmark in place.

Reply:
Or maybe `Resolve`'s logic that does the transpose and broadcasting on the array could be exposed as a method? 🤔

Reply:
On closer inspection that would be rather complicated, because `_as_compatible_cubes`, which does the transpose and broadcast, is [sometimes called a second time](https://github.com/SciTools/iris/blob/1e2a85ae14dc4552dae860d6079d157b39d1a9ca/lib/iris/common/resolve.py#L1210-L1225).  I think that also means my existing method is unsafe, so I have gone for a version of your suggestion.  It does not seem to make a difference to my shiny new benchmark:


```
Benchmarks that have stayed the same:

| Change   | Before [045ce020] <pearsonr-resolve~1>   | After [83d2c832] <pearsonr-resolve>   | Ratio   | Benchmark (Parameter)                |
|----------|------------------------------------------|---------------------------------------|---------|--------------------------------------|
|          | 615±300μs                                | 606±100μs                             | 0.99    | import_iris.Iris.time_analysis_stats |
|          | 76.6±10ms                                | 73.1±8ms                              | 0.95    | stats.PearsonR.time_lazy             |
|          | 37.1±6ms                                 | 30.3±5ms                              | ~0.82   | stats.PearsonR.time_real             |
```